modin/pandas/indexing.py

```diff
@@ -689,49 +781,70 @@ def _compute_lookup(self, row_loc, col_loc):
+            if isinstance(axis_loc, slice) or is_range_like(axis_loc):
+                if isinstance(axis_loc, slice) and axis_loc == slice(None):
+                    axis_lookup = axis_loc
+                else:
+                    axis_lookup = self.qc.get_axis(axis).slice_indexer(
+                        axis_loc.start, axis_loc.stop, axis_loc.step
+                    )
+                    axis_lookup = pandas.RangeIndex(
+                        axis_lookup.start,
+                        axis_lookup.stop,
+                        axis_lookup.step,
+                    )
+            elif self.qc.has_multiindex(axis):
+                # `Index.get_locs` raises an IndexError by itself if missing labels were provided,
+                # we don't have to do missing-check for the received `axis_lookup`.
+                if isinstance(axis_loc, pandas.MultiIndex):
+                    axis_lookup = self.qc.get_axis(axis).get_indexer_for(axis_loc)
+                else:
+                    axis_lookup = self.qc.get_axis(axis).get_locs(axis_loc)
+            elif is_boolean_array(axis_loc):
+                axis_lookup = boolean_mask_to_numeric(axis_loc)
             else:
-                col_lookup = self.qc.columns.get_locs(col_loc)
-        elif is_boolean_array(col_loc):
-            # If passed in a list of booleans, we return the index of the true values
-            col_lookup = [i for i, col_val in enumerate(col_loc) if col_val]
-        else:
-            col_lookup = self.qc.columns.get_indexer_for(col_loc)
-        return row_lookup, col_lookup
+                if is_list_like(axis_loc) and not isinstance(
+                    axis_loc, (np.ndarray, pandas.Index)
+                ):
+                    # `Index.get_indexer_for` works much faster with numpy arrays than with python lists,
+                    # so although we lose some time here on converting to numpy, `Index.get_indexer_for`
+                    # speedup covers the loss that we gain here.
+                    axis_loc = np.array(axis_loc)
+                axis_lookup = self.qc.get_axis(axis).get_indexer_for(axis_loc)
+                # `Index.get_indexer_for` sets -1 value for missing labels, we have to verify whether
+                # there are any -1 in the received indexer to raise a KeyError here.
+                missing_mask = axis_lookup < 0
+                if missing_mask.any():
+                    missing_labels = (
+                        # Converting `axis_loc` to maskable `np.array` to not fail
+                        # on masking non-maskable list-like
+                        np.array(axis_loc)[missing_mask]
+                        if is_list_like(axis_loc)
+                        else axis_loc
+                    )
+                    raise KeyError(f"Missing labels were provided: {missing_labels}")
+
+            if isinstance(axis_lookup, pandas.Index) and not is_range_like(axis_lookup):
+                axis_lookup = axis_lookup.values
```

Comment:
Some of the operations with `pandas.Index` returns a `pandas.Index`, for example:
```python
>>> pandas.Index([1, 2, 3])[[0, 1]]
Int64Index([1, 2], dtype='int64')
```
It seems that currently, there is no way of `axis_lookup` being a `pandas.Index`, however for future changes, this check would not allow a `pandas.Index` object to be leaked from this function.

Reply:
but it would allow range-like `pandas.Index` leak?.. what's the point?

Reply:
The function's return type is: `slice(None) if full axis grab, pandas.RangeIndex if repetition is detected, numpy.ndarray otherwise`, that means we don't want to return any `pandas.Index` except for `pandas.RangeIndex`, this check ensures it.